source/loader/layers/sanitizer/asan_shadow.cpp

```diff
@@ -0,0 +1,270 @@
+    }
+
+    uptr ShadowBegin = MemToShadow(Ptr);
+    uptr ShadowEnd = MemToShadow(Ptr + Size - 1);
+    assert(ShadowBegin <= ShadowEnd);
+    getContext()->logger.debug(
+        "EnqueuePoisonShadow(addr={}, count={}, value={})", (void *)ShadowBegin,
+        ShadowEnd - ShadowBegin + 1, (void *)(size_t)Value);
+    memset((void *)ShadowBegin, Value, ShadowEnd - ShadowBegin + 1);
+
+    return UR_RESULT_SUCCESS;
+}
+
+ur_result_t ShadowMemoryGPU::Setup() {
+    // Currently, Level-Zero doesn't create independent VAs for each contexts, if we reserve
+    // shadow memory for each contexts, this will cause out-of-resource error when user uses
+    // multiple contexts. Therefore, we just create one shadow memory here.
+    static ur_result_t Result = [this]() {
+        size_t ShadowSize = GetShadowSize();
+        // TODO: Protect Bad Zone
+        auto Result = getContext()->urDdiTable.VirtualMem.pfnReserve(
+            Context, nullptr, ShadowSize, (void **)&ShadowBegin);
+        if (Result == UR_RESULT_SUCCESS) {
+            ShadowEnd = ShadowBegin + ShadowSize;
+            // Retain the context which reserves shadow memory
+            getContext()->urDdiTable.Context.pfnRetain(Context);
+        }
+
+        // Set shadow memory for null pointer
+        ManagedQueue Queue(Context, Device);
+
+        Result = EnqueuePoisonShadow(Queue, 0, 1, kNullPointerRedzoneMagic);
+        if (Result != UR_RESULT_SUCCESS) {
+            getContext()->logger.error("EnqueuePoisonShadow(NullPointerRZ): {}",
+                                       Result);
+            return Result;
+        }
+        return Result;
+    }();
+    return Result;
+}
+
+ur_result_t ShadowMemoryGPU::Destory() {
+    if (ShadowBegin == 0) {
+        return UR_RESULT_SUCCESS;
+    }
+    static ur_result_t Result = [this]() {
+        auto Result = getContext()->urDdiTable.VirtualMem.pfnFree(
+            Context, (const void *)ShadowBegin, GetShadowSize());
+        getContext()->urDdiTable.Context.pfnRelease(Context);
+        return Result;
+    }();
+    return Result;
+}
+
+ur_result_t ShadowMemoryGPU::EnqueuePoisonShadow(ur_queue_handle_t Queue,
+                                                 uptr Ptr, uptr Size,
+                                                 u8 Value) {
+    if (Size == 0) {
+        return UR_RESULT_SUCCESS;
+    }
+
+    uptr ShadowBegin = MemToShadow(Ptr);
+    uptr ShadowEnd = MemToShadow(Ptr + Size - 1);
+    assert(ShadowBegin <= ShadowEnd);
+    {
+        static const size_t PageSize =
+            GetVirtualMemGranularity(Context, Device);
+
+        ur_physical_mem_properties_t Desc{
+            UR_STRUCTURE_TYPE_PHYSICAL_MEM_PROPERTIES, nullptr, 0};
+
+        // Make sure [Ptr, Ptr + Size] is mapped to physical memory
+        for (auto MappedPtr = RoundDownTo(ShadowBegin, PageSize);
+             MappedPtr <= ShadowEnd; MappedPtr += PageSize) {
+            std::scoped_lock<ur_mutex> Guard(VirtualMemMapsMutex);
```

Comment:
IMO, the lock range is too large here, it'd be better to lock the operations used on VirtualMemMaps only.